### PR TITLE
Introducing MemoryLimit

### PR DESCRIPTION
**Pull request by giorgiosironi**
If the process reaches too much allocated memory, it terminates and is
respawned to free it.

Requiring elife/bus-sdk has some conflict, so I'll try to solve that tomorrow.

### CODE REVIEW

File: src/modules/jcms_notifications/jcms_notifications.services.yml [hunk 1]
```diff
@@ -9,4 +9,8 @@ services:
   jcms_notifications.notification_storage:
     class: Drupal\jcms_notifications\MysqlNotificationStorage
     arguments: ['@database']
+  jcms_notifications.limit_service:
+    class: eLife\Bus\Limit\MemoryLimit
+    factory: ['eLife\Bus\Limit\MemoryLimit', 'mb']
+    arguments: 128
```
Comment (jamiehollern) [inline src/modules/jcms_notifications/jcms_notifications.services.yml]: This should be `[arguments: 128]`, which is why the build is failing. Once this is changed the build should pass.

Comment (jamiehollern) [inline src/modules/jcms_notifications/jcms_notifications.services.yml]: In addition, it should work as an integer but if it doesn't you mean need to pass `128` as a string.

Comment (thewilkybarkid) [inline src/modules/jcms_notifications/jcms_notifications.services.yml]: `arguments: [128]`?

Comment (jamiehollern) [inline src/modules/jcms_notifications/jcms_notifications.services.yml]: @thewilkybarkid Yes, that one... I'll update my comment!

**Comment by giorgiosironi** [comment]
Updated the arguments, but the problem was also related to adding the composer dependency.

**Comment by giorgiosironi** [comment]
Here is the actual problem:
```
$ composer update
Gathering patches for root package.
Removing package drupal/entityqueue so that it can be re-installed and re-patched.
Deleting web/modules/contrib/entityqueue - deleted
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Conclusion: don't install symfony/serializer v2.8.17
    - Conclusion: don't install drupal/core 8.2.6
    - Conclusion: remove symfony/serializer v2.8.16
    - Conclusion: don't install symfony/serializer v2.8.16
    - Conclusion: don't install symfony/serializer v2.8.15
    - Conclusion: don't install symfony/serializer v2.8.14
    - Conclusion: remove drupal/core 8.2.5
    - Conclusion: don't install symfony/serializer v2.8.13
    - Conclusion: don't install symfony/serializer v2.8.12
    - Conclusion: don't install symfony/serializer v2.8.11
    - Conclusion: don't install symfony/serializer v2.8.10
    - Conclusion: don't install symfony/serializer v2.8.9
    - Conclusion: don't install drupal/core 8.2.5
    - Conclusion: don't install symfony/serializer v2.8.8
    - Conclusion: don't install symfony/serializer v2.8.7
    - Conclusion: don't install symfony/serializer v2.8.6
    - Conclusion: don't install symfony/serializer v2.8.5
    - Conclusion: don't install drupal/core 8.2.4
    - Conclusion: don't install drupal/core 8.2.3
    - Conclusion: don't install drupal/core 8.2.2
    - Conclusion: don't install drupal/core 8.2.1
    - Conclusion: don't install drupal/core 8.2.0
    - Conclusion: don't install drupal/core 8.2.0-rc2
    - Conclusion: don't install drupal/core 8.2.0-rc1
    - Conclusion: don't install symfony/serializer v2.8.4
    - Conclusion: don't install drupal/core 8.3.0-beta1
    - Conclusion: don't install symfony/serializer v2.8.3
    - Conclusion: don't install drupal/core 8.2.0-beta3
    - Conclusion: don't install symfony/serializer v2.8.2
    - Conclusion: don't install drupal/core 8.2.0-beta2
    - Conclusion: don't install symfony/serializer v2.8.1
    - Conclusion: don't install drupal/core 8.2.0-beta1
    - Conclusion: don't install symfony/serializer v2.8.0
    - Conclusion: don't install drupal/core 8.3.0-alpha1
    - Installation request for elife/bus-sdk dev-master -> satisfiable by elife/bus-sdk[dev-master].
    - Conclusion: don't install symfony/serializer v2.8.0-BETA1
    - drupal/core 8.4.x-dev requires symfony/serializer ~2.8 -> satisfiable by symfony/serializer[v2.8.16, 2.8.x-dev, v2.8.0, v2.8.0-BETA1, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.17, v2.8.2, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - drupal/core 8.3.x-dev requires symfony/serializer ~2.8 -> satisfiable by symfony/serializer[v2.8.16, 2.8.x-dev, v2.8.0, v2.8.0-BETA1, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.17, v2.8.2, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - drupal/core 8.2.x-dev requires symfony/serializer ~2.8 -> satisfiable by symfony/serializer[v2.8.16, 2.8.x-dev, v2.8.0, v2.8.0-BETA1, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.17, v2.8.2, v2.8.3, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9].
    - Conclusion: don't install symfony/serializer 2.8.x-dev
    - Installation request for drupal/core ^8.2 -> satisfiable by drupal/core[8.2.0, 8.2.0-beta1, 8.2.0-beta2, 8.2.0-beta3, 8.2.0-rc1, 8.2.0-rc2, 8.2.1, 8.2.2, 8.2.3, 8.2.4, 8.2.5, 8.2.6, 8.2.x-dev, 8.3.0-alpha1, 8.3.0-beta1, 8.3.x-dev, 8.4.x-dev].
```
There is something in `elife/bus-sdk` that conflicts with the rest of Drupal. We could probably make those dependencies optional in `elife/bus-sdk`, but which one is the problem? I suspect https://github.com/elifesciences/bus-sdk-php/blob/master/composer.json#L11 but not sure so I would like to avoid removing random packages until it works.

**Comment by thewilkybarkid** [comment]
Looks like the `elife/api-sdk` and `symfony/serializer`.

**Comment by thewilkybarkid** [comment]
Symfony 3 is not going to hit Drupal until 8.4 (https://www.drupal.org/node/2712647), so we'll need to refactor the SDK to allow for `"symfony/serializer": "^2.8 || ^3.0"`.

**Comment by thewilkybarkid** [comment]
I've opened https://github.com/elifesciences/api-sdk-php/pull/113.

**Comment by jamiehollern** [comment]
@giorgiosironi @nlisgo One of the Travis builds has failed but the other has passed. It won't allow me to restart it, though.

**Comment by thewilkybarkid** [comment]
Shouldn't this PR have an updated `composer.lock`? (Are we really doing `composer update` on deploy?)

**Comment by giorgiosironi** [comment]
It should, but previously Composer refused to update

**Comment by thewilkybarkid** [comment]
Errr... of course. /me goes back to being quiet.

**Comment by giorgiosironi** [comment]
Looks like we have some configuration being carried over between builds, as the fresh site install fails:
```
You have requested a non-existent service "cache.backend.redis". Did
15:15:22 [184.72.214.138] out:                   you mean one of these: "cache.backend.apcu", "cache.backend.php",
15:15:22 [184.72.214.138] out:                   "cache.backend.memory"
```
We should merge first the Redis changed and then rebase this.

**Comment by jamiehollern** [comment]
@giorgiosironi Strangely, the Redis one fails also. Has the formula been updated on the boxes?

**Comment by giorgiosironi** [comment]
Logged in and check the `config/local.settings.php`, was up to date, let's focus on #95

**Comment by giorgiosironi** [comment]
Current status:
```
15:49:08 [54.86.132.53] out:           ID: composer-install
15:49:08 [54.86.132.53] out:     Function: cmd.run
15:49:08 [54.86.132.53] out:         Name: composer --no-interaction install
15:49:08 [54.86.132.53] out:       Result: False
15:49:08 [54.86.132.53] out:      Comment: Command "composer --no-interaction install" run
15:49:08 [54.86.132.53] out:      Started: 15:47:47.480782
15:49:08 [54.86.132.53] out:     Duration: 2740.15 ms
15:49:08 [54.86.132.53] out:      Changes:   
15:49:08 [54.86.132.53] out:               ----------
15:49:08 [54.86.132.53] out:               pid:
15:49:08 [54.86.132.53] out:                   10862
15:49:08 [54.86.132.53] out:               retcode:
15:49:08 [54.86.132.53] out:                   1
15:49:08 [54.86.132.53] out:               stderr:
15:49:08 [54.86.132.53] out:                   Loading composer repositories with package information
15:49:08 [54.86.132.53] out:                   Installing dependencies (including require-dev) from lock file
15:49:08 [54.86.132.53] out:                   Package operations: 12 installs, 37 updates, 2 removals
15:49:08 [54.86.132.53] out:                     - Removing drupal/console-extend-plugin (0.4.0)
15:49:08 [54.86.132.53] out:                     - Installing ocramius/package-versions (1.1.2) Loading from cache
15:49:08 [54.86.132.53] out:                   
15:49:08 [54.86.132.53] out:                                                       
15:49:08 [54.86.132.53] out:                     [InvalidArgumentException]        
15:49:08 [54.86.132.53] out:                     The "" directory does not exist.  
15:49:08 [54.86.132.53] out:                                                       
15:49:08 [54.86.132.53] out:                   
15:49:08 [54.86.132.53] out:                   install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--] [<packages>]...
15:49:08 [54.86.132.53] out:               stdout:
15:49:08 [54.86.132.53] out:                   Gathering patches for root package.
15:49:08 [54.86.132.53] out:                   Deleting web/modules/contrib/redis - deleted
15:49:08 [54.86.132.53] out:                   Gathering patches for root package.
15:49:08 [54.86.132.53] out:                   Gathering patches for dependencies. This might take a minute.
```

**Comment by giorgiosironi** [comment]
Rebased... merging when green.

**Pull request merged**